### DISCUSSION TITLE
Running native application on host

### DISCUSSION THREAD

**Discussion by tmirza** [eclipse-theia/theia#10992] [Q&A]
Hi,
I am new to Eclipse theia and still exploring different features of it. I have a question about browser app. I have an openocd like GDB server that I want to run on host-side (where browser is running) to provide debug functionality over board connected with my host machine via USB.
Is it possible to implement debug launch mechanism in a way that I can run this native application on user's host machine and then my backend/server side connect to it and provide debugging functionality? I see that there is a concept of front-end application contribution. Can it be used for this?

**Comment by vince-fugnitto** [comment]
@tmirza welcome to the community :+1: 

I definitely believe it is possible, in fact it seems like something that [**keil studio cloud**](https://www.keil.arm.com/) (developed by [Arm](https://www.arm.com/) and built using the Eclipse Theia framework) supports as a use-case.

Here is a debugging example video they reference on their homepage regarding using the cloud version of the IDE and being able to debug a board connected by USB on the machine:
- https://www.keil.arm.com/static/img/web-enabled-debugging.b071d66c087b.mp4

Many different use-cases should be possible, and if something isn't the framework is flexible enough to the point you can extend or re-implement parts to suit your needs. It is quite an involving use-case however and will likely require you to dig deeper into the framework.

**Reply by tmirza** [reply]
Thanks @vince-fugnitto 

Yes, I saw this demo while I was exploring things. I am exploring framework and will share if I find something. Meanwhile, if you have any pointers, kindly share with me.

**Reply by tmirza** [reply]
@vince-fugnitto,

I looked at different demos of KeilStudio. It looks to me that they use [dapjs](https://github.com/ARMmbed/dapjs) via WebUSB to provide debug support over cmsis-dap for Arm MCUs. I couldn't find any information if they support OpenOCD or not. However, while looking at Arduino Pro IDE, they seem to be running their agent locally on host by asking user to download it and then run. Their Theia based IDE then uses it.
Also, I spent some time exploring Debug Adapters and how those can be used. I extended CDT GDB adapter for some of my changes while exploring it. However, what I have seen so far is that my debug adapter and debugger (GDB in my case) both run on remote host (Server) and my GDB expects to connect to GDB server over TCP. This TCP thing itsself is a problem in case of web based IDEs.
Is there any recommendation to move such an environment to browser based IDE using Theia? Even If I ignore Open OCD or any GDB server running on host for now and consider using something like dapjs that I implement in frontend plugin, how can my debugger (GDB) connect to it without TCP?
